File: daily/2024/1455_Check_If_a_Word_Occurs_As_a_Prefix_of_Any_Word_in_a_Sentence.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int isPrefixOfWord(string sentence, string searchWord) {
        int n = sentence.length();
        int m = searchWord.length();
        int idx = 0;
        int ptr = 0;
        int ans = 0;
        while(idx < n){
            ans++;
            while(sentence[idx] == searchWord[ptr]){
                idx++, ptr++;
                if (ptr == m) return ans;
            }
            ptr = 0;
            while(idx < n && sentence[idx] != ' ') idx++;
            idx++;
        }
        return -1;
    }
};

int main(){
    string sentence = "i love eating burger";
    string searchWord = "burg";
    // string sentence = "this problem is an easy problem";
    // string searchWord = "pro";
    // string sentence = "i am tired";
    // string searchWord = "you";
    Solution S;
    cout << S.isPrefixOfWord(sentence, searchWord) << endl;
    return 0;
}

File: daily_python/1455_Check_If_a_Word_Occurs_As_a_Prefix_of_Any_Word_in_a_Sentence.py
import os
import sys
from typing import List

class Solution:
    def isPrefixOfWord(self, sentence: str, searchWord: str) -> int:
        n = len(sentence)
        m = len(searchWord)
        idx = ptr = 0
        ans = 0
        while idx < n:
            ans += 1
            while sentence[idx] == searchWord[ptr]:
                idx += 1
                ptr += 1
                if ptr == m:
                    return ans
            ptr = 0
            while idx < n and sentence[idx] != ' ':
                idx += 1
            idx += 1
        return -1
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            sentence = lines[0].split('=')[1].strip()[1:-2]
            searchWord = lines[1].split('=')[1].strip()[1:-2]
            dataset.append((sentence, searchWord))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for sentence, searchWord in dataset:
        results.append(solution.isPrefixOfWord(sentence, searchWord))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')